bidict/_base.py

```diff
@@ -496,15 +497,15 @@ def _init_from(self, other: MapOrIterItems[KT, VT]) -> None:
     def __or__(self: BT, other: t.Mapping[KT, VT]) -> BT:
         """Return self|other."""
         if not isinstance(other, t.Mapping):
-            return NotImplemented
+            return NotImplemented  # flagged by pyright
         new = self.copy()
         new._update(other, rbof=False)
         return new
 
     def __ror__(self: BT, other: t.Mapping[KT, VT]) -> BT:
         """Return other|self."""
         if not isinstance(other, t.Mapping):
-            return NotImplemented
+            return NotImplemented  # flagged by pyright
```

Comment:
This seems like a pyright bug to me. I suspect there's special handling for NotImplemented but it doesn't handle the interaction with TypeVars.

Another workaround is `cast(Any, NotImplemented)`.

Reply:
Indeed, thanks, @JelleZijlstra! (Much appreciate your help with this PR, didn't even realize you'd notice it!)

FWIW, I think that using something like `# pyright: ignore [link to bug]` is preferable to `cast(Any, NotImplemented)` here, because:
1. It makes it clear that this is only necessary to work around a bug in pyright
2. It clearly does not affect how mypy is interpreting this code

Reply:
`NotImplemented` is defined in a strange way in typeshed. The definition (which I've pasted below) alludes to the fact that "it makes the error message better". I presume this is referring to the error message emitted by mypy. It's probably not a good idea to work around mypy-specific error handling by adding hacks like this in typeshed. But I digress.

```python
@final
class _NotImplementedType(Any):  # type: ignore[misc]
    # A little weird, but typing the __call__ as NotImplemented makes the error message
    # for NotImplemented() much better
    __call__: NotImplemented  # type: ignore[valid-type]

NotImplemented: _NotImplementedType
```

As Jelle conjectured, the reason you're seeing different behavior between `__eq__` and `__ror__` is because the former returns `bool` and the latter returns a TypeVar. Pyright is treating those two inconsistently. `_NotImplementedType` inherits from `Any`, so it should be allowed in the TypeVar case.

I've filed the following pyright bug to track this: https://github.com/microsoft/pyright/issues/3171

Reply:
Your take on typeshed here makes sense to me. I hope it can be improved in the future.

Thanks for following up on this with microsoft/pyright#3171! I'll keep on eye on that issue. In the meantime, does this help make the case for supporting something like
```python
return NotImplemented  # pyright: ignore  # see microsoft/pyright#3171
```
?

It seems like this would provide users hitting this or any future bugs with a more forgiving and maintainable escape hatch. (Especially if, once the bug were fixed, the next version of pyright could emit some kind of "`unused "pyright: ignore"` error à la mypy.)

Reply:
You can use `# type: ignore` if you don't want to fix it. There is no such thing as `# mypy: ignore`, and it likewise doesn't make sense to have a `# pyright: ignore`.

Reply:
No, it doesn't make sense to add `# pyright: ignore`, just like it wouldn't make sense for you to request that mypy add a `# mypy: ignore`. There's already a standardized way to tell type checkers that you want to suppress errors: `# type: ignore`.

Reply:
That's true. But mypy supports `# type: ignore [mypy_rule]`, which somewhat mitigates this, whereas pyright does not.

Note that I keep saying "*something like* `# pyright: ignore`". By this I mean something that would accomplish the same thing, not necessarily that exact syntax. And the syntax need not be decided at the same time that "*some* syntax should support this" be decided. But in case it's helpful, just to throw out another syntax idea, rules could be prefixed: `# type: ignore [mypy:<mypy_rule>,pyright:<pyright_rule>]` (I know this is not standardized yet, but meanwhile we need to get stuff done.)

Reply:
If you want, you can use `# type: ignore [pyright]` or `# type: ignore [reportGeneralTypeIssue]` or `# type: ignore -- custom comment explaining why you are ignoring the type checker`. All of these work today with pyright.

Reply:
But then mypy will fail with an “unused type: ignore” error. Is there a solution that works today for pyright that does not cause mypy to start failing? It seems that improved support might be needed by both type checkers to enable more interoperability for “type: ignore” comments that are meant for one and not the other type checker. (That could be achieved with the prefix syntax I proposed as one example.)

Reply:
(mypy could also add support for inline comments like `mypy: disable`, which would let me change the `type: ignore [mypy_rule]` comments I’m currently using on lines that only mypy and not pyright is flagging.)

Reply:
You don't need to enable the "unused type: ignore" feature. That's an optional feature that doesn't really make sense when you're running multiple type checkers on the same code base.

Reply:
IIRC I’m not explicitly turning that on. I’m only explicitly turning on mypy strict mode. Not sure if I can turn it off without turning off strict mode. But as I mentioned in a recent comment, I’d prefer to keep this on. That way, when a bug I worked around is fixed in a new version, I get promoted to remove the now unused suppression.

Reply:
It looks like I can add `warn_unused_ignores = False` (which gets implicitly set to True by the `strict = True` at the beginning), to the end of [my mypy.ini](https://github.com/jab/bidict/blob/main/mypy.ini), to silence all mypy output about any unused `type: ignore` comments, without having to give up any of the other desired checks that `strict = True` enables.

Minimized example:
```console
❯ cat test.py
x = 1 + 1  # type: ignore

❯ mypy test.py
Success: no issues found in 1 source file

❯ mypy --strict test.py  # exits nonzero
test.py:1: error: Unused "type: ignore" comment
Found 1 error in 1 file (checked 1 source file)

❯ mypy --strict --no-warn-unused-ignores test.py  # exits zero
Success: no issues found in 1 source file
```
I don't see any way to run mypy in strict mode and have it *warn, but not error* on unused type: ignores, in other words still exit zero, but also still output a warning message for any unused typed: ignores. @JelleZijlstra, am I missing something? If not, shouldn't that be supported?

As https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-unused-ignores says?

> This flag, along with the [--warn-redundant-casts](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-redundant-casts) flag, are both particularly useful when you are upgrading mypy. Previously, you may have needed to add casts or # type: ignore annotations to work around bugs in mypy or missing stubs for 3rd party libraries.
>
> These two flags let you discover cases where either workarounds are no longer necessary.

Quite validating to see this in the docs (hadn't noticed it before), as it matches my exact experience.

So to summarize all my current suggestions:
- Both mypy and pyright should support a mode where they complain in their output about any inline comments intended to suppress an error that are not actually suppressing an error.
  - It should be possible to configure whether this results in an error (nonzero exit status), a warning (exit zero, but still complain in the output), or total silence (exit zero, no mention in the output).
- Both mypy and pyright should support targeted inline comments that allow users to unambiguously suppress an error that only one of them is emitting (possibly erroneously due to a bug), without affecting the behavior of the other one.

Reply:
After upgrading to pyright 1.1.229, pyright is no longer giving false positives here. Nice.

I'll leave this thread unresolved in case @JelleZijlstra can give thoughts on adding the following to mypy (and link any newly-created issues if interested):
* support for warning-but-not-erroring on unused suppression comments
* support for "mypy: ignore" (and not just "type: ignore") suppression comments
  * If this were available, I could replace all my "type: ignore" comments with "mypy: ignore" comments in the case that only mypy and not pyright is flagging the associated lines (which I can easily see thanks to pyright's `reportUnnecessaryTypeIgnoreComment=warning`).
  * Then for each "mypy: ignore" suppression, decide whether it's a pyright false negative or a mypy false positive, and ensure there is a corresponding open bug I can link to.
  * Finally, that will leave all remaining "type: ignore" comments indicating lines that pyright, mypy, and I all agree should be type errors, but where turning off the suppression would require typing features I can't use yet, or would make the code less maintainable, or where I just need to move on for now and come back to the particular line when I have time.

Reply:
I see the use case, but I think mypy doesn't currently have a "warning" output level, and mypy already has a fiendishly complicated set of configuration options.

Reply:
Fair enough!

That was the first bullet. What about the second (supporting “mypy: ignore”)?